### ISSUE TITLE
Couldn't connect to SMTP host - Bad Request

### ISSUE THREAD

**Issue by DasNerdwork** [mailcow/mailcow-dockerized#2549]
Hey,

I configured mailcow to run on my FQDN mail.dasnerdwork.net and use custom ports 82 for http and 444 for https. Could log into the user interface on the http port but got a 400 bad request from nginx on the https port. 
Tried to connect my webservers wordpress SMTP with mailcow with the following settings:

SMTP Host: mail.dasnerdwork.net
Type: TLS
SMTP Port: 587
And my already created SMTP user admin@dasnerdwork.net with it's password.

This is the following output I receive from testing the SMTP connection.

**SMTP Debug:**
_2019-04-21 23:36:34	Connection: opening to v22018056297666752.ultrasrv.de:82, timeout=10, options=array ()
2019-04-21 23:36:34	Connection: opened
2019-04-21 23:36:44	SERVER -> CLIENT:
2019-04-21 23:36:44	CLIENT -> SERVER: EHLO dasnerdwork.net
2019-04-21 23:36:44	SERVER -> CLIENT: HTTP/1.1 400 Bad Request
Server: nginx
2019-04-22 00:00:34	SMTP ERROR: EHLO command failed: HTTP/1.1 400 Bad Request
2019-04-22 00:00:34	SMTP NOTICE: EOF caught while checking if connected
2019-04-22 00:00:34	Connection: closed
2019-04-22 00:00:34	SMTP Error: Could not connect to SMTP host.
2019-04-22 00:00:34	SMTP connect() failed._

In the first place I'm already happy that I could make it run on the subdomain on http and aswell send as receive mails, but all the wordpress mails should be sent as admin@dasnerdwork.net which sadly doesn't work without a proper SMTP connection.

I literally beg you for help, as I'm trying to fix this since several days already....
Thank you in advance

**Comment by junelsolis** [comment]
> 2019-04-21 23:36:34	Connection: opening to v22018056297666752.ultrasrv.de:82, timeout=10, options=array ()

Hello there,

You said your FQDN is mail.dasnerdwork.net. Is there any reason why the first line in your SMTP debug says `v22018056297666752.ultrasrv.de:82`? Have you setup reverse DNS?

The other problem that I seem to notice is that you mentioned that port 82 is for HTTP. Shouldn't this connection be made to port 25 / 465 / 587 on mail.dasnerdwork.net?

**Comment by andryyy** [comment]
I also see "opening to v22018056297666752.ultrasrv.de:82" => Port 82.

Why does it connect to port 82 instead of 587? There might be some NAT going on.

**Comment by DasNerdwork** [comment]
The thing is, I have no idea why. I set my FQDN on the Debian Server to mail.dasnerdwork.net and the PTR also to the same. The v22018... is only the name of the server at my Hoster. I don’t have it written in any configs so I got no clue where that comes from. Still if you only type the v22018...de into the browser it redirects to my homepage DasNerdwork.net

As the webinterface didn’t work on port 80 and 443 on mail.dasnerdwork.net (it redirects me to DasNerdwork.net) I tried to change its port to 82 and 444, which resulted in the webinterface being reachable through mail.dasnerdwork.net:82, as well as dasnerdwork.net:82 and dasnerdwork.net:444, but NOT mail.dasnerdwork.net:444 which results in nginx 400 Bad request.

Tried everything from dasnerdwork.net with every combination of ports and without and also mail.dasnerdwork.net, nothing worked yet

**Comment by junelsolis** [comment]
> As the webinterface didn’t work on port 80 and 443 on mail.dasnerdwork.net (it redirects me to DasNerdwork.net)

When I browse to `mail.dasnerdwork.net` I get the Apache2 Debian Default Page. Perhaps the reason why you couldn't get the MC web interface to work on ports 80 and 443 is that they're already bound by Apache 2 on your server.

And could you please tell us if your Wordpress site is running on the same server as Mailcow?

**Comment by DasNerdwork** [comment]
Yes it is. I wanted to try to let the Mailserver run on the subdomain and wordpress on the main domain, but both in the same server. I can send the rest of my configuration in ~30-40 minutes if you need/want to see it.

**Comment by junelsolis** [comment]
> I can send the rest of my configuration in ~30-40 minutes if you need/want to see it.

I think that would be helpful.

I think there should be no problem having the mailcow-nginx container listen on ports 82 and 444. If you're absolutely sure that Wordpress is set to connect to port 587, then maybe @andryyy is right that there might be some NAT going on.

Maybe you can try setting Wordpress to connect to `127.0.0.1` instead of `mail.dasnerdwork.net`? This is, of course, assuming that Wordpress is not also running as a Docker container.

**Comment by DasNerdwork** [comment]
Just completely removed any RDNS configs and stuff like that in nginx and apache sites available,  fully removed mailcow and reinstalled it. Now the interface is only reachable through mail.dasnerdwork.net:82, dasnerdwork.net:82 and dasnerdwork.net:444 , but I 400 Bad Request on the subdomains https 444.

SkipsLetsEncrypt is set to "n"

![mailcowconf](https://user-images.githubusercontent.com/49847819/56848991-ee3d2300-68ee-11e9-9342-0cb1ad9caca1.png)

![Hostnames](https://user-images.githubusercontent.com/49847819/56848992-f006e680-68ee-11e9-848b-213cb6e69c4f.png)

**Comment by junelsolis** [comment]
Could you please try running `telnet 127.0.0.1 587` and post a screenshot of the output?

**Comment by DasNerdwork** [comment]
I edited the first sentence of my last comment, made a mistake, just so you know

**Comment by andryyy** [comment]
You should get a support package or find someone to properly set it up via remote hands. 

I don’t know why you bind HTTPS to port 82. 
Or how your reverse proxy is configured that it acts like this. 

It will also never obtain a certificate when it is configured to listen on 82/444, when your reverse proxy is not configured to forward Let’s Encrypt requests to your cow. 

You should try to understand the reverse proxy a little bit more before proceeding. :) No offense at all! It just seems a little bit messed up.

**Comment by DasNerdwork** [comment]
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
220 mail.dasnerdwork.net ESMTP Postcow

**Comment by junelsolis** [comment]
> Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
220 mail.dasnerdwork.net ESMTP Postcow

So have you tried setting Wordpress to connect to `127.0.0.1` instead of `mail.dasnerdwork.net` yet?

**Comment by DasNerdwork** [comment]
No, I'll try now.

**Comment by DasNerdwork** [comment]
Also I already tried with configuring Let's Encrypt to forwards to those ports but I couldn't manage it to work as I never did that before and had struggles with that aswell. Don't really need a reverse proxy though as I don't even understand why every tutorial I find recommends using one. I read some texts about it, but is it really necessary for a little own mailserver where the goal is to send mails from wordpress as admin@dasnerdwork.net and redirect all incoming mail traffic to a private mail?

**Comment by junelsolis** [comment]
> Also I already tried with configuring Let's Encrypt to forwards to those ports but I couldn't manage it to work as I never did that before and had struggles with that aswell.

As @andryyy said above, the only way Mailcow (or the acme-mailcow) container will be able to obtain  LetsEncrypt certificates is if it can bind to port 80. Right now, it can't because that's occupied by your Apache server. Your other options are to setup a reverse proxy for the autoconfig.* server_name or to have some sort of backend certbot work that places the certificate files where both your web server AND Mailcow can use them.

> Don't really need a reverse proxy though as I don't even understand why every tutorial I find recommends using one

There's probably a very good reason why most tutorials you have found recommend using a reverse proxy. I think that a reverse proxy is the best way to go with your specific setup. In case you haven't come across it yet, this [link](https://mailcow.github.io/mailcow-dockerized-docs/firststeps-rp/) in the documentation is a good starting point.

**Comment by DasNerdwork** [comment]
If the wordpress website runs on Apache the proxy should run on nginx right?
Or does that matter in this case?

**Comment by DasNerdwork** [comment]
When I tried to setup the reverse proxy I did after [this](https://felixmoessbauer.com/blog-reader/mailcow-reverse-proxy-letsencrypt.html) explanation. However it did not work after changing every mx.<your-domain>.de to mail.dasnerdwork.net, even though the mxtoolbox says at least the DNS settings for SMTP 
![dns](https://user-images.githubusercontent.com/49847819/56849491-6c9cc380-68f5-11e9-81e9-17d86ac8094f.png)
are setup correctly.

**Comment by DasNerdwork** [comment]
Do you recommend any literature for reverse proxy or some good videos about it so I can get a better understanding of how it works?

**Comment by junelsolis** [comment]
I think there are some concepts you have not completely understood yet. For example, your reverse proxy configuration only deals with serving websites -- it should have nothing to do with how email is sent and delivered through your server (SMTP).

There are some good examples and documentation on how to configure an NGINX reverse proxy at [nginx.com](https://www.nginx.com) and [nginx.org](https://www.nginx.org)

**Comment by DasNerdwork** [comment]
Ok, I did understand the use and need of a proxy now. Basically using different processes (like here mailcow and wordpress) on the same port. 

I used the [mailcow documentation](https://mailcow.github.io/mailcow-dockerized-docs/firststeps-rp/) to, edited "CHANGE_TO_MAILCOW_HOSTNAME" to mail.dasnerdwork.net and copied the cert and key created with certbot to /opt/mailcow-dockerized/data/assets/ssl/

I think it should work. I can't see anything missing or wrong, did not get any errors too. But now it still redirects me from mail.dasnerdwork.net to the main page dasnerdwork.net

This is probably some misconfiguration somewhere else, but I can't find it. Does one of you have any idea where this problem occurs?

This is my .htaccess in /var/www/html/wordpress
```
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteBase /
RewriteRule ^index\.php$ - [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule . /index.php [L]
</IfModule>
```

This is my nginx configuration as a symlink in sites-enabled from sites-available

```
server {
  listen 80 default_server;
  listen [::]:80 default_server;
  server_name mail.dasnerdwork.net autodiscover.*;
  return 301 https://$host$request_uri;
}
server {
  listen 80;
  listen [::]:80;
  server_name autoconfig.*;
  rewrite ^/(.*)$ /autoconfig.php last;
  location / {
    proxy_pass http://127.0.0.1:8080/;
    proxy_set_header Host $http_host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    client_max_body_size 0;
  }
}
server {
  listen 443;
  server_name mail.dasnerdwork.net autodiscover.* autoconfig.*;

  ssl on;
  ssl_certificate /opt/mailcow-dockerized/data/assets/ssl/cert.pem;
  ssl_certificate_key /opt/mailcow-dockerized/data/assets/ssl/key.pem;
  ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
  ssl_ciphers HIGH:!aNULL:!MD5;

  location / {
      proxy_pass http://127.0.0.1:8080/;
      proxy_set_header Host $http_host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      client_max_body_size 0;
  }
}
```

**Comment by amorfo77** [comment]
Hi @DasNerdwork , i created a gist from my config, you can check [here](https://gist.github.com/amorfo77/e9d2a08cf76bf2cbaecce89c765f4763) 
now i'm running mailcow behind a reverse proxy, also the proxy serves 2 wordpress sites and 1 gitea app.

Hope it helps.

**Comment by DasNerdwork** [comment]
Yes @amorfo77, that helped a lot. Thank you.

I managed to get it working now. I migrated wordpress from apache2 to nginx and added both the mailcow aswell as the wordpress configs in nginx/sites-available with symlinks to site-enable, reinstalled the certbot certificates with certbot --nginx and copied them to mailcows preferred location.

mail.dasnerdwork.net runs now the mailserver properly with https
dasnerdwork.net runs the wordpress website with https
and the SMTP connection is reachable on mail.dasnerdwork.net:587 TLS.

Thanks again to all of you for helping me and your little advices :)
